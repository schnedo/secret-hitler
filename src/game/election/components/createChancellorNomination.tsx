import type { ComponentType, ReactElement } from "react";
import styled from "styled-components";
import type { ModalProps } from "../../../components";
import type { Player, PlayerId } from "../../player";
import type { AvatarProps } from "../../player/components";
import type Government from "../Government";

const NominationRow = styled.div`
  display: flex;
`;

export interface NominationValidator {
  (
    numPlayers: number,
    lastGovernment: Government | null,
    nomination: Government,
  ): boolean;
}

export interface ChancellorNominationProps {
  onNomination: (playerId: PlayerId) => void;
  nominationValidator: NominationValidator;
  players: Player[];
  presidentialCandidate: number | null;
  government: Government | null;
}

export default function createChancellorNomination(
  Modal: ComponentType<ModalProps>,
  Avatar: ComponentType<AvatarProps>,
) {
  return function ChancellorNomination({
    onNomination,
    nominationValidator,
    players,
    presidentialCandidate,
    government,
  }: ChancellorNominationProps): ReactElement {
    const electablePlayers =
      presidentialCandidate === null
        ? []
        : players.filter((player, id) =>
            nominationValidator(players.length, government, {
              president: presidentialCandidate,
              chancellor: id,
            }),
          );

    return (
      <Modal open>
        <div>Please nominate your chancellor candidate.</div>
        {electablePlayers.map((player) => (
          <NominationRow>
            <button
              onClick={() =>
                onNomination(players.findIndex((pl) => pl === player))
              }
            >
              Nominate {player.name}
            </button>
            <Avatar player={player} />
          </NominationRow>
        ))}
      </Modal>
    );
  };
}
