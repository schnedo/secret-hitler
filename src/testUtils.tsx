import type { PropsWithChildren } from "react";

export function MockComponent({
  children,
  ...props
}: PropsWithChildren<unknown>): JSX.Element {
  return (
    <div>
      {JSON.stringify(props)} {children}
    </div>
  );
}
