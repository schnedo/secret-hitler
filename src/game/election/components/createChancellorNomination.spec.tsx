import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockComponent } from "../../../testUtils";
import type { Player } from "../../player";
import type Government from "../Government";
import createChancellorNomination, {
  NominationValidator,
} from "./createChancellorNomination";

const ChancellorNomination = createChancellorNomination(
  MockComponent,
  // @ts-expect-error probably some type inference issue. can be ignored here, as it is a mock component
  MockComponent,
);
const somePlayers: Player[] = [
  {
    name: "John",
    role: "fascist",
    title: null,
  },
  {
    name: "Martha",
    role: "hitler",
    title: null,
  },
  {
    name: "Bob",
    role: "liberal",
    title: null,
  },
  {
    name: "Alice",
    role: "liberal",
    title: null,
  },
  {
    name: "Mohammed",
    role: "liberal",
    title: null,
  },
];
const presidentialCandidate = 1;
const government: Government = { chancellor: 3, president: 4 };

it("should render correctly", async () => {
  expect.hasAssertions();

  const { container } = render(
    <ChancellorNomination
      onNomination={() => undefined}
      nominationValidator={() => true}
      players={somePlayers}
      presidentialCandidate={presidentialCandidate}
      government={government}
    />,
  );

  expect(container).toMatchSnapshot();
});

it("should call onNomination with correct player id", async () => {
  expect.hasAssertions();

  const handleNomination = jest.fn();
  render(
    <ChancellorNomination
      onNomination={handleNomination}
      nominationValidator={() => true}
      players={somePlayers}
      presidentialCandidate={presidentialCandidate}
      government={government}
    />,
  );

  await Promise.all(
  somePlayers.map(async ({ name }, index) => {
    await userEvent.click(screen.getByRole("button", { name: `Nominate ${name}` }));
    expect(handleNomination).toHaveBeenLastCalledWith(index);
  }));
});

it("should only show players that are electable for nomination", async () => {
  expect.hasAssertions();

  const handleNomination = jest.fn();
  const validPlayers = [0, 2];
  const validateNomination = jest
    .fn<ReturnType<NominationValidator>, Parameters<NominationValidator>>()
    .mockImplementation((numPlayers, lastGovernment, nomination) =>
      validPlayers.includes(nomination.chancellor),
    );
  render(
    <ChancellorNomination
      onNomination={handleNomination}
      nominationValidator={validateNomination}
      players={somePlayers}
      presidentialCandidate={presidentialCandidate}
      government={government}
    />,
  );

  const allNominationButtons = screen.getAllByRole("button");
  expect(allNominationButtons).toHaveLength(validPlayers.length);
  allNominationButtons.forEach((nominationButton, index) => {
    expect(nominationButton).toHaveTextContent(
      somePlayers[validPlayers[index]].name,
    );
  });
});

it("should not show presidentialCandidate as an option", async () => {
  expect.hasAssertions();

  render(
    <ChancellorNomination
      onNomination={() => true}
      nominationValidator={() => true}
      players={somePlayers}
      presidentialCandidate={presidentialCandidate}
      government={government}
    />,
  );

  expect(
    screen.queryByRole("button", {
      name: somePlayers[presidentialCandidate].name,
    }),
  ).not.toBeInTheDocument();
});
