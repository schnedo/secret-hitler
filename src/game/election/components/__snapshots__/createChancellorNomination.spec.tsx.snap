// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div>
    {"open":true}
     
    <div>
      Please nominate your chancellor candidate.
    </div>
    <div
      class="c0"
    >
      <button>
        Nominate 
        John
      </button>
      <div>
        {"player":{"name":"John","role":"fascist","title":null}}
         
      </div>
    </div>
    <div
      class="c0"
    >
      <button>
        Nominate 
        Martha
      </button>
      <div>
        {"player":{"name":"Martha","role":"hitler","title":null}}
         
      </div>
    </div>
    <div
      class="c0"
    >
      <button>
        Nominate 
        Bob
      </button>
      <div>
        {"player":{"name":"Bob","role":"liberal","title":null}}
         
      </div>
    </div>
    <div
      class="c0"
    >
      <button>
        Nominate 
        Alice
      </button>
      <div>
        {"player":{"name":"Alice","role":"liberal","title":null}}
         
      </div>
    </div>
    <div
      class="c0"
    >
      <button>
        Nominate 
        Mohammed
      </button>
      <div>
        {"player":{"name":"Mohammed","role":"liberal","title":null}}
         
      </div>
    </div>
  </div>
</div>
`;
